logic/odmr_logic.py

```diff
@@ -454,41 +489,64 @@ def mw_sweep_on(self):
 
         limits = self.get_hw_constraints()
         param_dict = {}
-
+        self.frequency_lists = []
+        self.final_freq_list = []
         if self.mw_scanmode == MicrowaveMode.LIST:
-            if np.abs(self.mw_stop - self.mw_start) / self.mw_step >= limits.list_maxentries:
-                self.log.warning('Number of frequency steps too large for microwave device. '
-                                 'Lowering resolution to fit the maximum length.')
-                self.mw_step = np.abs(self.mw_stop - self.mw_start) / (limits.list_maxentries - 1)
-                self.sigParameterUpdated.emit({'mw_step': self.mw_step})
-
-            # adjust the end frequency in order to have an integer multiple of step size
-            # The master module (i.e. GUI) will be notified about the changed end frequency
-            num_steps = int(np.rint((self.mw_stop - self.mw_start) / self.mw_step))
-            end_freq = self.mw_start + num_steps * self.mw_step
-            freq_list = np.linspace(self.mw_start, end_freq, num_steps + 1)
-            freq_list, self.sweep_mw_power, mode = self._mw_device.set_list(freq_list,
+            final_freq_list = []
+            used_starts = []
+            used_steps = []
+            used_stops = []
+            for mw_start, mw_stop, mw_step in zip(self.mw_starts, self.mw_stops, self.mw_steps):
+                num_steps = int(np.rint((mw_stop - mw_start) / mw_step))
+                end_freq = mw_start + num_steps * mw_step
+                freq_list = np.linspace(mw_start, end_freq, num_steps + 1)
+
+                # adjust the end frequency in order to have an integer multiple of step size
+                # The master module (i.e. GUI) will be notified about the changed end frequency
+                final_freq_list.extend(freq_list)
+                # self.frequency_lists.append(freq_list)
+
+                used_starts.append(mw_start)
+                used_steps.append(mw_step)
+                used_stops.append(end_freq)
+
+            final_freq_list = np.array(final_freq_list)
+            if len(final_freq_list) >= limits.list_maxentries:
+                self.log.error('Number of frequency steps too large for microwave device.')
```

Comment:
After the error message you have to properly leave the function. The following code (same as at the end of the function) will do the job:
`mode, is_running = self._mw_device.get_status()`
`self.sigOutputStateUpdated.emit(mode, is_running)`
`return mode, is_running`